ethtx/decoders/abi/helpers/utils.py

```diff
@@ -0,0 +1,119 @@
+#  Copyright 2021 DAI Foundation
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at: http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import logging
+from typing import Iterator, Optional, List
+
+from ethtx.models.semantics_model import (
+    FunctionSemantics,
+    ParameterSemantics,
+    Signature,
+    SignatureArg,
+)
+from ethtx.providers import FourByteProvider
+from ethtx.providers.semantic_providers.semantics_repository import SemanticsRepository
+from ethtx.providers.signature_provider import SignatureProvider
+
+log = logging.getLogger(__name__)
+
+
+def decode_function_abi_with_external_source(
+    signature: str,
+    repository: SemanticsRepository,
+    _provider: Optional[SignatureProvider] = FourByteProvider,
+) -> Iterator[FunctionSemantics]:
+    function = repository.get_most_used_signature(signature_hash=signature)
+    if function:
+        log.info(
+            "Successfully guessed function from SemanticsRepository - %s.",
+            function.json(),
+        )
+        function_semantics = FunctionSemantics(
+            signature,
+            function.name,
+            _prepare_parameter_semantics(function.args, function.tuple, unknown=False),
+            [],
+        )
+        yield function_semantics
+        return
```

Comment:
is this return needed if result is yielded? Or shouldn't yield be return?

Reply:
Yes, it's needed. It prevents us from moving on and we keep function as an iterator (needed in ABI checks). After that, we got value from yield and `return` triggers `StopIteration`.